src/fs/dir_entry_iter.rs

```diff
@@ -0,0 +1,35 @@
+use crate::fs::file_info::{BoxedUefiFileInfo, UEFI_FILE_INFO_MAX_SIZE};
+use super::*;
+use alloc_api::vec::Vec;
+
+/// Iterator over the low-level entries of a UEFI directory handle.
+/// Returns owning [BoxedUefiFileInfo].
+pub struct UefiDirectoryEntryIterator {
+    handle: UefiDirectoryHandle,
+    buffer: Vec<u8>,
+}
+
+impl UefiDirectoryEntryIterator {
+    pub fn new(handle: UefiDirectoryHandle) -> Self {
+        // May contain 256 characters with an fixed size UTF-16 format. Null-byte included.
+        let mut buffer = Vec::with_capacity(UEFI_FILE_INFO_MAX_SIZE);
+        (0..buffer.capacity()).for_each(|_| buffer.push(0));
+        Self { handle, buffer }
+    }
+}
+
+impl Iterator for UefiDirectoryEntryIterator {
+    type Item = BoxedUefiFileInfo;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // reset as the buffer gets used multiple times
+        self.buffer.fill(0);
+
+        self.handle
+            .read_entry(&mut self.buffer)
+            // it is extremly ugly to marry the results from uefi with the option from the iterator..
+            //  thus, just panic =(
+            .expect("Uefi Error")
```

Comment:
Wouldn't it be an idea to have an iterator which returns `Result<BoxedUefiFileInfo>`, so the users can just call `collect` if they want to get a `Result<Vec<BoxedUefiFileInfo>>`?

Reply:
ah, that's one option. Is it a thing that iterators return results?

Reply:
See [`std::fs::read_dir`](https://doc.rust-lang.org/std/fs/fn.read_dir.html) for an example of an iterator that returns results (specifically `next()` returns `Option<io::Result<fs::DirEntry>`).